client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+
+### min\_Rt 
+
+The Transaction maintains the minimum read time it has seen so far.  On every read, if the read time is less than min\_Rt, we update min\_Rt.  
+
+Note that this read time is NOT the time the Transaction read the entry.  Transaction reads from cache do not update entry read times; cache reads from the DB server do update entry read times. 
+
+### max\_Vt
+
+The Transaction maintains the maximum value time it has seen so far.  On every read, if the value time is greater than max\_Vt, we update max\_Vt.
+
+Note that updating max\_Vt can cause the invariant violation described above, causing the Transaction read method to throw StaleException(Vt).  Intuitively, when reading an entry from the DB, we have no idea when it was last updated until we actually read it.  As a result, we cannot avoid the possibility of violating our consistent snapshot invariant.
+
+* **Question (6):** If the user fails to specify max\_age correctly, should we throw a StaleException?
+
+### view 
+
+The view maintains the set of operations performed throughout the Transaction.  When the user commits the transactions, each updated value in the view is sent to the cache, which writes the changes through to the DB server.  (If the view's values are out-of-date with the DB, write will throw a StaleException.)
+
+Like the cache, the view is structured as a ("table:key" map) of (sorted value maps).  The "table:key" map has keys composed of the concatenation of (table identifier) and (key) strings; its values are the (sorted value maps) described next.  The (sorted value maps) are from (LONG\_MAX - read_timestamp) keys to values (JSON), as in the cache.  (Note, as mentioned above, the keys are (LONG\_MAX - read\_timestamp) rather than just (read\_timestamp) because we want the table's entries to be sorted in reverse, from most recent to oldest timestamp.)  
+
+Additionally, every "table:key" key in the view also stores two booleans called "updated" and "deleted" alongside its sorted map.  The "updated" boolean is False if the Transaction has only read the value of that key and True if the Transaction has written an update to that key's value.  The "deleted" boolean is True if and only if the user has called the delete method on its key during this Transaction.
+
+Upon commit, we search through the map keys for updated values (updated==True) and deleted values (deleted==True) and construct a set of op\_lists from the most recent value (sorted by (LONG\_MAX - read\_timestamps)) in each updated key's sorted map.  Note the op\_list is a parameter to Cache's write method, and each table needs its own op\_list.
```

Comment:
Everything in the view becomes part of the batch. The values that are read only become `hold` ops. Even though it doesn't update them, the Tx depends on them not having changed. I recommend using `op` as `create`, `hold`, `update`, and `delete` rather than multiple boolean flags, for sanity.